### PR TITLE
feat: forbid redundant schema definitions

### PR DESCRIPTION
**Pull request by fbiville**

### CODE REVIEW

**Comment by neo4j-connectors** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>PR Title 'Forbid redundant schema definitions': subject may not be empty, type may not be empty</td>
    </tr>
  </tbody>
</table>



> Pull request title should follow [Conventional commits](https://cheatography.com/albelop/cheat-sheets/conventional-commits).
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against bd73cd55a6ec0f63ea135281a02280065085668c
</p>

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 1]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                .isInstanceOf(InvalidSpecificationException.class)
+                .hasMessageContainingAll(
+                        "2 error(s)",
+                        "0 warning(s)",
+                        "$.targets.nodes[0].schema.key_constraints[0].properties[0] \"not-a-prop\" is not part of the property mappings",
+                        "$.targets.nodes[0].schema.unique_constraints[0].properties[0] \"not-a-prop\" is not part of the property mappings");
+    }
+
+    @Test
+    public void fails_if_key_constraint_range_index_are_defined_on_same_labels_and_properties() {
```
Comment (fbiville) [inline src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java:7886]:
```suggestion
    public void fails_if_key_constraint_and_range_index_are_defined_on_same_labels_and_properties() {
```

**Comment by neo4j-connectors** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>:exclamation: Please include a description of your PR changes.</td>
    </tr>
  </tbody>
</table>


> Pull request should have a description of the underlying changes.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 97858681e7ab70f180da68a43a13fa1f7ad20111
</p>

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 2]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                .isInstanceOf(InvalidSpecificationException.class)
+                .hasMessageContainingAll(
+                        "2 error(s)",
+                        "0 warning(s)",
+                        "$.targets.nodes[0].schema.key_constraints[0].properties[0] \"not-a-prop\" is not part of the property mappings",
+                        "$.targets.nodes[0].schema.range_indexes[0].properties[0] \"not-a-prop\" is not part of the property mappings");
+    }
+
+    @Test
+    public void fails_if_unique_constraint_range_index_are_defined_on_same_labels_and_properties() {
```
Comment (fbiville) [inline src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java:8173]:
```suggestion
    public void fails_if_unique_constraint_and_range_index_are_defined_on_same_labels_and_properties() {
```

**Review by fbiville — COMMENTED** [review]
_

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 3]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                }]
+            }
+        }
+        """
+                                .stripIndent())))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void does_not_fail_if_key_and_existence_constraints_are_not_defined_on_same_labels_and_properties() {
```
Comment (Emrehzl94) [inline src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java:7473]: nit: Maybe naming could be does_not_fail_if_key_and_existence_constraints_are_defined_on_same_labels_but_different_properties
Wdyt?

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 4]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                        }]
+                    }
+                }
+                """
+                                .stripIndent())))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void does_not_fail_if_key_and_unique_constraints_are_not_defined_on_same_labels_and_properties() {
```
Comment (Emrehzl94) [inline src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java:7759]: same here about naming

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 5]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                        }]
+                    }
+                }
+                """
+                                .stripIndent())))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void does_not_fail_if_unique_constraint_and_range_index_are_not_defined_on_same_labels_and_properties() {
```
Comment (Emrehzl94) [inline src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java:8334]: same here

**Review by Emrehzl94 — APPROVED** [review]
LGTM ✅ just a few comments about naming

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java (hunk 3 above)
Comment (fbiville) [reply src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java]: fixed in 6a76978eba15b54b7f35eaa33fd5f22afcb8ae2c

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java (hunk 4 above)
Comment (fbiville) [reply src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java]: fixed in 6a76978eba15b54b7f35eaa33fd5f22afcb8ae2c

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java [hunk 6]
```diff
@@ -7306,4 +7306,1154 @@ public void fails_if_node_schema_vector_index_options_is_wrongly_typed() {
...
+                        }]
+                    }
+                }
+                """
+                                .stripIndent())))
+                .doesNotThrowAnyException();
+    }
+
+    @Test
+    public void does_not_fail_if_key_constraint_and_range_index_are_not_defined_on_same_labels_and_properties() {
```
[the remark this answers is not in the document]
Comment (fbiville) [reply src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java]: fixed in 6a76978eba15b54b7f35eaa33fd5f22afcb8ae2c

File: src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java (hunk 5 above)
Comment (fbiville) [reply src/test/java/org/neo4j/importer/v1/ImportSpecificationDeserializerNodeTargetTest.java]: fixed in 6a76978eba15b54b7f35eaa33fd5f22afcb8ae2c

**Pull request merged**